lottery/src/lib.rs

```diff
@@ -0,0 +1,286 @@
+        } else {
+            panic!("leave_lottery(): Lottery on: {}", self.lottery_is_on());
+        }
+    }
+
+    /// Gets the player's balance
+    /// Requirements:
+    /// * Lottery has started and lottery time has not expired
+    /// * Player must be on the player list
+    /// Arguments:
+    /// * `index`: lottery player index
+    fn get_balance(&mut self, index: u32) {
+        if self.lottery_is_on() {
+            if let Some(player) = self.players.get(&index) {
+                msg::reply(Event::Balance(player.balance), 0);
+            } else {
+                panic!("get_balance(): Player {} not found", index);
+            }
+        } else {
+            panic!("get_balance(): Lottery on: {}", self.lottery_is_on());
+        }
+    }
+
+    /// Gets a list of players
+    /// Requirements:
+    /// * Lottery has started and lottery time has not expired
+    /// * List of players must not be empty
+    fn get_players(&mut self) {
+        if self.lottery_is_on() && !self.players.is_empty() {
+            msg::reply(Event::Players(self.players.clone()), 0);
+        } else {
+            panic!(
+                "get_players(): Lottery on: {}  players.is_empty(): {}",
+                self.lottery_is_on(),
+                self.players.is_empty()
+            );
+        }
+    }
+
+    // Random number generation from block timestamp
+    fn get_random_number(&mut self) -> u32 {
+        let timestamp: u64 = exec::block_timestamp();
+        let code_hash: sp_core::H256 = blake2_256(&timestamp.to_be_bytes()).into();
+        let u_buf = code_hash.to_fixed_bytes();
+        let mut number: u32 = 0;
+
+        for u in u_buf {
+            number += u as u32;
+        }
+
+        number
+    }
+
+    /// Lottery winner calculation
+    /// Requirements:
+    /// * Only owner can pick the winner
+    /// * Lottery has started and lottery time is expired
+    /// * List of players must not be empty
+    fn pick_winner(&mut self) {
+        if msg::source() == self.lottery_owner
+            && self.lottery_state.lottery_started
+            && self.lottery_state.lottery_start_time + self.lottery_state.lottery_duration
+                <= exec::block_timestamp()
+            && !self.players.is_empty()
+        {
+            let index: u32 = self.get_random_number() % (self.players.len() as u32);
+
+            if let Some(win_player) = self.players.get(&index) {
+                msg::send_bytes(win_player.player_id, b"Winner", exec::value_available());
+                self.lottery_history
+                    .insert(self.lottery_id, win_player.player_id);
+                msg::reply(Event::Winner(index), 0);
+            } else {
+                panic!("pick_winner(): Player {} not found", index);
+            }
+
+            debug!("Winner: {}", index);
+
+            self.lottery_state = LotteryState::default();
+            self.players = BTreeMap::new();
+            self.players_timestamp = BTreeMap::new();
```

Comment:
May we keep timestamp instead of `Player` instead of creating a separate structure for it?